### PR TITLE
Core/Entities: Creature hover v3

### LINKED ISSUE #15177: [3.3.5] Hover Mode not working (55$ bounty)
In the air phase of Mimiron, he does not remain floating. After a few seconds of battle it ceases to hover and keep attacking on the floor.

Mimiron (Airphase)
wowscrnshot_070815_125924

> bug caused by incomplete hover movement implementation
> 
> current hover movement implementation works for static npcs only, they start moving and hover position goes to ground
> 
> on movement target destination is relocated to ground, hover offset is missing

rev b63082971fac8e239f08db0d16113c46c7c1e3ee

### PR DESCRIPTION
**Pull request by Shauren**
**Changes proposed:**
Hover shit

**Target branch(es):** 3.3.5/master

- [X] 3.3.5
- [X] master

**Issues addressed:**
Closes #15177

**Tests performed:** (Does it build, tested in-game, etc.)
Mimiron looks good

Analyzing sniff data it turns out that Z coord is ground + hoverheight

### CODE REVIEW

**Comment by ccrs** [comment]
what about the other movegens?

**Comment by ccrs** [comment]
calling ````UpdateAllowedPositionZ```` on each position relocation..., is there a way around this?

**Comment by Shauren** [comment]
Other movegens use .MoveByPath feeded by PathGenerator which already calls UpdateAllowedPositionZ on every point

**Comment by sirikfoll** [comment]
Still no effect on Valkyr Shadowguard, at Charge(74445) use, she drops to ground position, sniffs show that she shouldn't get bellow GroundLevel + HoverHeight
Should be fixed in `void Spell::EffectChargeDest(SpellEffIndex /*effIndex*/)` or try to correct dest by SpellScript?
<details> Packets
ServerToClient: SMSG_ON_MONSTER_MOVE (0xAA6D) Length: 50 ConnIdx: 0 Time: 06/25/2011 21:47:55.000 Number: 418338
GUID: Full: 0xF1508F010001622A Type: Vehicle Entry: 36609 (Val'kyr Shadowguard) Low: 90666
Toggle AnimTierInTrans: false
Position: X: 507.9648 Y: -2137.583 Z: 853.3589
Move Ticks: 3094058
Spline Type: 0 (Normal)
Spline Flags: 8192 (Flying)
Move Time: 615
Waypoints: 1
[0] Waypoint: X: 531.8093 Y: -2152.725 Z: 844.8569

ServerToClient: SMSG_SPELL_GO (0xEC2D) Length: 49 ConnIdx: 0 Time: 06/25/2011 21:47:55.000 Number: 418340
Caster GUID: Full: 0xF1508F010001622A Type: Vehicle Entry: 36609 (Val'kyr Shadowguard) Low: 90666
Caster Unit GUID: Full: 0xF1508F010001622A Type: Vehicle Entry: 36609 (Val'kyr Shadowguard) Low: 90666
Cast Count: 0
Spell ID: 74399 (Charge)
Cast Flags: 768 (Unknown7, Unknown8)
Time: 3361219318
Hit Count: 0
Miss Count: 0
Target Flags: 64 (DestinationLocation)
Destination Transport GUID: 0x0
Destination Position: X: 531.8093 Y: -2152.725 Z: 840.8569
Unk Byte 2: 1

**Comment by Shauren** [comment]
Val'kyr Shadowguard is not hovering on TC but uses DisableGravity (will need fixing)

**Comment by killerwife** [comment]
Hm. Based from my experience with nightbane, the following occurs:
Nightbane plays liftoff emote, SetHover(true), SetLevitate(true) (SMSG_SPLINE_MOVE_GRAVITY_DISABLE is the name i think), UNIT_BYTE1_FLAG_FLY_ANIM. When these are combined nightbane flies nicely in place without changing Z in any way. Is Mimiron different from this?

**Comment by Shauren** [comment]
Mimiron does this - spawn high up in the air (inside that metal thingy) with MOVEMENTFLAG_DISABLE_GRAVITY movement flag, float slowly down to groundZ + UNIT_FIELD_HOVERHEIGHT then removes MOVEMENTFLAG_DISABLE_GRAVITY and enables MOVEMENTFLAG_HOVER

**Comment by killerwife** [comment]
Curious, but then again TBC doesn't have UNIT_FIELD_HOVERHEIGHT, so not too surprising that there is a difference.

**Comment by joschiwald** [comment]
brundir in ulduar is also working well, but there is now one problem with spell target selection with spell 61884
WorldObject::MovePositionToFirstCollision should ignore hoverheight when checking for big z changes, i made following change and it is working well
```diff
diff --git a/src/server/game/Entities/Object/Object.cpp b/src/server/game/Entities/Object/Object.cpp
index 9302dc18ea..850bfb2c58 100644
--- a/src/server/game/Entities/Object/Object.cpp
+++ b/src/server/game/Entities/Object/Object.cpp
@@ -2306,7 +2306,7 @@ void WorldObject::MovePositionToFirstCollision(Position &pos, float dist, float
     for (uint8 j = 0; j < 10; ++j)
     {
         // do not allow too big z changes
-        if (std::fabs(pos.m_positionZ - destz) > 6.0f)
+        if (std::fabs(pos.m_positionZ - destz) > (6.0f + (isType(TYPEMASK_UNIT) ? ToUnit()->GetHoverOffset() : 0.0f)))
         {
             destx -= step * std::cos(angle);
             desty -= step * std::sin(angle);
```

**Comment by joschiwald** [comment]
another problem is killing hovering npcs, they stay in the air, and are not falling to ground

**Comment by lachtanek** [comment]
That seems easily fixed by setting hover height to 0 on death.

**Comment by joschiwald** [comment]
no it is not so easy like you think, i tried that in brundir script, and it was not working for some reason

**Comment by lachtanek** [comment]
Probably because setDeathState is called before AI::JustDied (same reason you can't set loot mode there, which is pretty annoying tbh), needs to be put before Creature.cpp:1863

**Comment by joschiwald** [comment]
@lachtanek disable hover movement on death must work without any script class (about the loot thing, changing loot mode on death is a bad approach)

**Comment by lachtanek** [comment]
I know, but as I understood you, you tried it in JustDied, which is called too late to work, if not, then my bad

**Comment by ccrs** [comment]
yay!
maybe separate flight into canfly and disablegravity?

**Comment by Shauren** [comment]
@ccrs ahead of you, thats already included in that commit

**Comment by ccrs** [comment]
```c++
enum class CreatureFlightMovementType : uint8
 {
     None,
     DisableGravity,
     CanFly,
 
     Max
 };
```
ooooh I see, I thought it was just a 1 - 0 value  👍

**Comment by Keader** [comment]
Testing this pr with #21031 (added HoverMode in constructor of boss_twin_baseAI)
And both valkyr looks like "walking falling in the air", like without this pr

File: sql/updates/world/3.3.5/2017_xx_xx_20961_world.sql [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+CREATE TABLE `creature_movement_override` (
+  `SpawnId` int(10) unsigned NOT NULL DEFAULT '0',
+  `Ground` tinyint(3) unsigned NOT NULL DEFAULT '1',
+  `Swim` tinyint(3) unsigned NOT NULL DEFAULT '1',
+  `Flight` tinyint(3) unsigned NOT NULL DEFAULT '0',
+  `Rooted` tinyint(3) unsigned NOT NULL DEFAULT '0',
+  PRIMARY KEY (`SpawnId`)
+) ENGINE=MyISAM DEFAULT CHARSET=utf8;
+
+INSERT INTO `creature_template_movement` SELECT `entry`,0,0,0,0 FROM `creature_template` WHERE `InhabitType`!=3;
```
Comment (Keader) [inline sql/updates/world/3.3.5/2017_xx_xx_20961_world.sql]: is not possible execute this sql 2 times
Error Code: 1054. Unknown column 'InhabitType' in 'where clause'

Comment (tkrokli) [reply sql/updates/world/3.3.5/2017_xx_xx_20961_world.sql]: Maybe parentheses `( )` is needed around the `SELECT` statement?

Comment (Shauren) [reply sql/updates/world/3.3.5/2017_xx_xx_20961_world.sql]: No sql which adds/removes columns can be executed twice

**Comment by Shauren** [comment]
@Keader tested it as well and it worked after I fixed database data.
* `creature_template_movement`.`Ground` = 2 for all difficulty entries
* `creature_template_addon`.`bytes1` = 0x2000000 for all difficulty entries
* `creature_template`.`HoverHeight` = 6 for all difficulty entries (currently only 10 normal has proper value)

File: src/server/game/Globals/ObjectMgr.cpp [hunk 2]
```diff
@@ -502,20 +503,31 @@ void ObjectMgr::LoadCreatureTemplate(Field* fields)
...
-    creatureTemplate.ModExperience  = fields[68].GetFloat();
-    creatureTemplate.RacialLeader   = fields[69].GetBool();
-
-    creatureTemplate.movementId         = fields[70].GetUInt32();
-    creatureTemplate.RegenHealth        = fields[71].GetBool();
-    creatureTemplate.MechanicImmuneMask = fields[72].GetUInt32();
-    creatureTemplate.flags_extra        = fields[73].GetUInt32();
-    creatureTemplate.ScriptID           = GetScriptId(fields[74].GetString());
+    if (!fields[62].IsNull())
+        creatureTemplate.Movement.Ground = static_cast<CreatureGroundMovementType>(fields[62].GetUInt8());
```
Comment (ariel-) [inline src/server/game/Globals/ObjectMgr.cpp]: Validate values?

**Comment by Keader** [comment]
need fix conflicts shauren :)

**Comment by Keader** [comment]
@Shauren Fix conflicts please

**Comment by tkrokli** [comment]
Looks like a rebase on main branch is needed.

File: src/server/game/Entities/Object/Object.cpp [hunk 3]
```diff
@@ -1463,66 +1459,44 @@ void WorldObject::UpdateAllowedPositionZ(float x, float y, float &z) const
...
-                        z = max_z;
-                    else if (z < ground_z)
-                        z = ground_z;
-                }
-            }
+            bool canSwim = unit->CanSwim();
+            float ground_z = z;
+            float max_z;
+            if (canSwim)
+                max_z = GetMap()->GetWaterOrGroundLevel(GetPhaseMask(), x, y, z, &ground_z, !unit->HasAuraType(SPELL_AURA_WATER_WALK));
```
Comment (Golrag) [inline src/server/game/Entities/Object/Object.cpp]: Why not GetMapWaterOrGroundLevel ? Missing since rebase ? :p

**Comment by Shauren** [comment]
this is going in after I fix merge conflicts

**Comment by bumbummen99** [comment]
I just saw that a claim has been made. It is nice to see that there is progress on this pr but could you provide some evidence that it is working? I currently have no time to setup a test env or to get a 3.3.5a client. If you can't and the other backers are okay with the claim i'm in.

**Comment by Shauren** [comment]
First, the queries against clean tdb
```
UPDATE `creature_template_addon` SET `bytes1`=0x2000000 WHERE `entry` IN (34496,34497);
INSERT INTO `creature_template_movement` (`CreatureId`,`Ground`,`Swim`,`Flight`,`Rooted`) VALUES
(34496,2,0,0,0), -- Ground = 2 means hover
(34497,2,0,0,0);
```

Short demo of valkyrs in Trial of the Crusader
https://youtu.be/DwT5hCz-v7g

**Comment by Seyden** [comment]
@Shauren what about master?

**Comment by Shauren** [comment]
master needs additional stuff on top of this

**Pull request merged**